packages/ts-client/src/webrtc/StateManager.ts

```diff
@@ -0,0 +1,548 @@
+    } catch (error) {
+      newTrack = {
+        ...newTrack,
+        metadata: undefined,
+        metadataParsingError: error,
+      };
+      trackContext.metadataParsingError = error;
+      trackContext.metadata = undefined;
+    }
+    newTrack = { ...newTrack, rawMetadata: trackMetadata };
+    trackContext.rawMetadata = trackMetadata;
+    endpoint.tracks.set(trackId, newTrack);
+
+    this.webrtc.emit('trackUpdated', trackContext);
+  }
+
+  public onTrackEncodingDisabled(data: any) {
+    if (this.getEndpointId() === data.endpointId) return;
+
+    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> =
+      this.idToEndpoint.get(data.endpointId)!;
+
+    if (endpoint == null)
+      throw `Endpoint with id: ${data.endpointId} doesn't exist`;
+
+    const trackId = data.trackId;
+    const encoding = data.encoding;
+
+    const trackContext = endpoint.tracks.get(trackId)!;
+
+    this.webrtc.emit('trackEncodingDisabled', trackContext, encoding);
+  }
+
+  public onTrackEncodingEnabled(data: any) {
+    if (this.getEndpointId() === data.endpointId) return;
+
+    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> =
+      this.idToEndpoint.get(data.endpointId)!;
+
+    if (endpoint == null)
+      throw `Endpoint with id: ${data.endpointId} doesn't exist`;
+
+    const trackId = data.trackId;
+    const encoding = data.encoding;
+
+    const trackContext = endpoint.tracks.get(trackId)!;
+
+    this.webrtc.emit('trackEncodingEnabled', trackContext, encoding);
+  }
+
+  public onEncodingSwitched(data: any) {
+    const trackId = data.trackId;
+    const trackContext = this.trackIdToTrack.get(trackId)!;
+    trackContext.encoding = data.encoding;
+    trackContext.encodingReason = data.reason;
+
+    trackContext.emit('encodingChanged', trackContext);
+  }
+
+  public onVadNotification(data: any) {
+    const trackId = data.trackId;
+    const ctx = this.trackIdToTrack.get(trackId)!;
+    const vadStatus = data.status;
+    if (isVadStatus(vadStatus)) {
+      ctx.vadStatus = vadStatus;
+      ctx.emit('voiceActivityChanged', ctx);
+    } else {
+      console.warn('Received unknown vad status: ', vadStatus);
+    }
+  }
+
+  public onBandwidthEstimation(data: any) {
+    this.bandwidthEstimation = data.estimation;
+
+    this.webrtc.emit('bandwidthEstimationChanged', this.bandwidthEstimation);
+  }
+
+  public validateAddTrack(
+    track: MediaStreamTrack,
+    simulcastConfig: SimulcastConfig,
+    maxBandwidth: TrackBandwidthLimit,
+  ): string | null {
+    const isUsedTrack = isTrackInUse(this.connection, track);
+
+    let error;
+    if (isUsedTrack) {
+      error =
+        "This track was already added to peerConnection, it can't be added again!";
+    }
+
+    if (!simulcastConfig.enabled && !(typeof maxBandwidth === 'number'))
+      error =
+        'Invalid type of `maxBandwidth` argument for a non-simulcast track, expected: number';
```

Comment:
Can this function return values, instead of storing them in variable?
```suggestion
      return 'Invalid type of `maxBandwidth` argument for a non-simulcast track, expected: number';
```